File: src/main/java/cpsc224/levels/Level.java
package cpsc224.levels;

import cpsc224.levels.rooms.Room;

public class Level {

    private final Room[][] rooms;
    private int numRooms;
    private int roomLength;

    private Coordinate startPosition;

    /**
     * Creates a level using the given initializer.
     * @param initializer the level initializer
     */
    public Level(LevelInitializer initializer) {
        rooms = initializer.initializeLevel();
        numRooms = initializer.getNumRooms() + 2;
        startPosition = initializer.getStartRoom();
        exploreRoom(startPosition);
    }

    /**
     * Gets the number of rooms in the level
     * @return the number of rooms
     */
    public int getNumRooms() {
        return numRooms;
    }

    /**
     * Gets the length of the level in rooms
     * @return the length
     */
    public int getRoomLength() {
        return rooms.length;
    }

    /**
     * Gets the 2D array of rooms
     * @return the rooms
     */
    public Room[][] getRooms() {
        return rooms;
    }

    /**
     * Gets the coordinate of the start room for the floor
     * @return the start coordinate
     */
    public Coordinate getStartPosition() {
        return startPosition;
    }

    /**
     * Gets a room at a specific coordinate
     * @param coord the coordinate of the room
     * @return the room at the coordinate
     */
    public Room getRoom(Coordinate coord) {
        return rooms[coord.getRow()][coord.getCol()];
    }

    /**
     * Sets the current position of the player. Updates what rooms are discovered and explored.
     * @param currentPosition the new position of the player
     */
    public void exploreRoom(Coordinate currentPosition) {
        int row = currentPosition.getRow();
        int col = currentPosition.getCol();

        if (rooms[row][col] != null)
            rooms[row][col].setExplored(true);

        if (row < rooms.length - 1 && rooms[row + 1][col] != null)
            rooms[row + 1][col].setDiscovered(true);
        if (row > 0 && rooms[row - 1][col] != null)
            rooms[row - 1][col].setDiscovered(true);
        if (col < rooms[row].length - 1 && rooms[row][col + 1] != null)
            rooms[row][col + 1].setDiscovered(true);
        if (col > 0 && rooms[row][col - 1] != null)
            rooms[row][col - 1].setDiscovered(true);
    }
}

File: web/src/game/levels/Level.ts
import { Coordinate } from './Coordinate';
import type { Room } from './rooms/Room';

export interface LevelInitializer {
  initializeLevel(): (Room | null)[][];
  getStartRoom(): Coordinate;
  getNumRooms(): number;
}

export class Level {
  private rooms: (Room | null)[][];
  private numRooms: number;
  private startPosition: Coordinate;
  private floorIndex: number;

  constructor(initializer: LevelInitializer, floorIndex = 0) {
    this.rooms = initializer.initializeLevel();
    this.numRooms = initializer.getNumRooms() + 2;
    this.startPosition = initializer.getStartRoom();
    this.floorIndex = floorIndex;
    this.exploreRoom(this.startPosition);
  }

  getFloorIndex(): number {
    return this.floorIndex;
  }

  getRooms(): (Room | null)[][] {
    return this.rooms;
  }

  getNumRooms(): number {
    return this.numRooms;
  }

  getStartPosition(): Coordinate {
    return this.startPosition;
  }

  getRoom(pos: Coordinate): Room | null {
    return this.rooms[pos.getRow()]?.[pos.getCol()] ?? null;
  }

  exploreRoom(pos: Coordinate): void {
    const room = this.getRoom(pos);
    if (!room) return;
    room.setExplored(true);

    const r = pos.getRow();
    const c = pos.getCol();
    const neighbors = [
      new Coordinate(r - 1, c),
      new Coordinate(r + 1, c),
      new Coordinate(r, c - 1),
      new Coordinate(r, c + 1),
    ];
    for (const n of neighbors) {
      const nr = this.getRoom(n);
      if (nr) nr.setDiscovered(true);
    }
  }
}
